core/src/raw/oio/write/two_ways_write.rs

```diff
@@ -0,0 +1,64 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use async_trait::async_trait;
+use bytes::Bytes;
+
+use crate::raw::oio::Streamer;
+use crate::raw::*;
+use crate::*;
+
+/// TwoWaysWrite is used to implement [`Write`] based on two ways.
+///
+/// Users can wrap two different writers together.
+pub enum TwoWaysWriter<L: oio::Write, R: oio::Write> {
```

Comment:
This part is a little bit confusing, I think our current use case is that two writers, one for normal writes and one for append writes. My suggestion would be just to rename the `Left` to `Writer` and `Right` to `AppendWriter`.

We could mark this as a todo though, this is not a big issue for this PR.

Reply:
They could be different write implementations, for example, gcs and azblob don't supports multipart uploads.

Anyway, we can improve this part later.